sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ShufflePartitionsUtil.scala

```diff
@@ -173,7 +176,11 @@ object ShufflePartitionsUtil extends Logging {
     if (numPartitions > start) {
       val partitionSpecs = coalescePartitions(
         partitionIndices(start), partitionIndices.last + 1, validMetrics, targetSize)
-      newPartitionSpecsSeq.zip(partitionSpecs).foreach(spec => spec._1 ++= spec._2)
+      assert(numPartitions > 0)
+      // skip empty partition iff all shuffle are empty after coalesced
+      if (!partitionSpecs.forall(specs => specs.size == 1 && specs.head.dataSize.contains(0))) {
```

Comment:
I looked into `coalescePartitions`, it does skip empty partitions, but it needs to at least return one `CoalescedPartitionSpec`.

I think we can pass one more boolean flag to `coalescePartitions` to allow it to return no `CoalescedPartitionSpec`, for coalescing partitions between skewed partitions.

Reply:
SGTM !